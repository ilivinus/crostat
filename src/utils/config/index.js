export const API_KEY = { 
    key_name : 'X-CMC_PRO_API_KEY', 
    value : '57440339-3174-489c-8cc3-54ed63a243df' 
}