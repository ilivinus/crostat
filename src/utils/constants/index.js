import { Platform } from 'react-native'

export const PAGE_SIZE = 20;
export const DEBOUNCE_INTERVAL = 2000;
export const colors = {
    red: '#F8333C',
    green: '#4ADF86',
    darkGray: '#31393C',
    greyOutline: '#bbbbbb',
    primary: '#00C0FF',
    primaryLight: 'rgba(0, 192, 255, 0.15)',
    lightGrey: '#B4B4B4',
    white: '#ffffff',
    transparent: 'transparent',
    darkBlue: '#161C36',
    black: '#000',
    selectionColor: Platform.OS === 'ios' ? '#00C0FF' : undefined
  };
  
  /**
   * App themes
   */
  export const themes = {
    dark: {
      ...colors,
      tabBarColor: '#192543',
      cardBackground: colors.darkBlue,
      headerTitleColor: colors.white,
      textColor: colors.white,
      thumbTintColor: colors.white,
    },
    light: {
      ...colors,
      tabBarColor: '#F3F3F3',
      cardBackground: colors.white,
      headerTitleColor: colors.darkGray,
      textColor: colors.darkGray,
      thumbTintColor: colors.darkGray,
    },
  };
  export const metrics = {
    //hitSlop: makeHitSlop,
    shadow: {
      basic: {
        shadowOffset: {
          width: 2,
          height: 0
        },
        shadowOpacity: 0.2,
        shadowColor: colors.black,
        shadowRadius: 6
      }
    }
  };
 export const properties = {
    feedbackEmail: {
      email: 'umeadumalivinus@gmail.com',
      subject: 'CloseCross App feedback',
    },
  };
  