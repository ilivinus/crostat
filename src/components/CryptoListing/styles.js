import { StyleSheet } from 'react-native';
import { colors } from '../../utils/constants';

const COIN_ICON_SIZE = 30;

const styles = StyleSheet.create({
    flatListRoot: {
        flex: 1,
    },
    separator: {
        backgroundColor: colors.transparent,
        height: 5,
    },
    contentContainerList: {
        alignSelf: 'stretch',
    },
    symbolName :{
        top: 5,
        position: 'absolute',        
        right: 10,
    },

    coinIcon: {
        height: COIN_ICON_SIZE,
        width: COIN_ICON_SIZE,
    },
    contentWrapper: {
        alignItems: 'center',
        flex: 1,
        justifyContent: 'center',
    },
    volumeText : {
        top: 5,
        left: 10,
        position: 'absolute',
    },
    timeText : {
        top: 5,
        right: 10,
        position: 'absolute',
    },
    holdingText: {
        bottom: 5,
        left: 10,
        position: 'absolute',
    },
    metaWrapper: {
        alignItems: 'center',
        flex: 0.4,
        justifyContent: 'center',
    },
    priceUsdText: {
        bottom: 5,
        position: 'absolute',
        right: 10,
    },
    root: {
        alignSelf: 'center',
        flexDirection: 'row',
        height: 130,
        padding: 5,
        position: 'relative',
        width: '100%',
    },
    title: {
        marginTop: 5,
        fontWeight: 'bold',
        textAlign: 'center',
    },
    titleWrapper: {
        alignItems: 'center',
        flex: 0.4,
        justifyContent: 'center',
    },
    totalWrapper: {
        alignItems: 'center',
        flexDirection: 'row',
    },
    noHolding: {
        color: colors.lightGrey,
    },
});

export default styles;