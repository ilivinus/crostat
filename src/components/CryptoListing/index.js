import React, { PureComponent } from 'react';
import { FlatList, RefreshControl, View } from 'react-native';
import { colors } from '../../utils/constants';
import styles from './styles';
import Spinner from '../commons/Loading'
const CryptoListing = (props) => (
  <View
    style={[styles.flatListRoot, { backgroundColor: props.screenProps.theme.tabBarColor }]}
  >
    {props.isLoading && props.listing.length == 0 
    ? (<Spinner />) : (<FlatList
      ItemSeparatorComponent={() => <View style={styles.separator} />}
      contentContainerStyle={styles.contentContainerList}
      data={props.listing}
      keyExtractor={props._keyExtractor}
      onEndReached={props._onEndReached}
      onEndReachedThreshold={0.5}
      refreshControl={
        <RefreshControl
          onRefresh={props._onRefresh}
          refreshing={props.refreshing}
          tintColor={colors.primary}
        />
      }
      renderItem={props._renderItem}
    />)}
  </View>

);

export default CryptoListing;