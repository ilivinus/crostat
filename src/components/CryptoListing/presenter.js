import React, { PureComponent } from 'react';
import { View, TouchableOpacity } from 'react-native';
import Ionicons from 'react-native-vector-icons/Ionicons';
import { Text, NameText } from '../commons/Typographie';
import { colors } from '../../utils/constants';
import { getIfPercentNegative } from '../../utils/helpers/getIfPercentNegative';
import { moneyThousand } from '../../utils/helpers/numbers';
import styles from './styles';
import PropType from 'prop-types';

class CryptoItemView extends PureComponent {
  _getIfNeg(percentChange) {
    const _percentChange = percentChange || '0';
    return getIfPercentNegative(_percentChange);
  }

  _getPercentChange(percent) {
    let str;
    const style = {};

    const color = this._getIfNeg(percent) ? colors.red : colors.green;
    
    if (percent == null) {
      str = '##';
      style.color = this.props.screenProps.theme.textColor;
    } else {
      str = `${parseFloat(percent).toFixed(2)}%`;
      style.color = color;
    }

    return <Text style={style}>{str}</Text>;
  }

  _getIconPercent(percent) {
    if (percent == null) {
      return null;
    }

    const props = {
      size: 25,
    };

    if (this._getIfNeg(percent)) {
      props.name = 'ios-trending-down';
      props.color = colors.red;
    } else {
      props.name = 'ios-trending-up';
      props.color = colors.green;
    }

    return <Ionicons {...props} />;
  }

  _getIconArrow(percent) {
    const props = {
      size: 25,
    };

    if (this._getIfNeg(percent)) {
      props.name = 'ios-arrow-round-down';
      props.color = colors.red;
    } else {
      props.name = 'ios-arrow-round-up';
      props.color = colors.green;
    }

    return <Ionicons  {...props} />;
  }

  _getPrice() {
    return moneyThousand(this.props.priceUsd);
  }

  _onNavigationPress = () => {
    this.props.navigation.navigate('CryptoDetailsScreen', {
      cryptoId: this.props.id,
      name: `${this.props.name} (${this.props.symbol})`,
    });
  };

  render() {
    if (this.props.name == null) {
      return null;
    }
    const { theme } = this.props.screenProps;

    return (
      <TouchableOpacity
        onPress={this._onNavigationPress}
        style={[styles.root, { backgroundColor: theme.cardBackground }]}
      >
      <Text style={[styles.volumeText, { color: theme.textColor }]}>
          <NameText style={[styles.volumeText, { color: colors.lightGrey }]}>
            Vol (24h)
          </NameText>{' '}
          {this.props.volume24h}{`(${this.props.convert})`}
        </Text>
        <View style={styles.titleWrapper}>
          <Text style={[styles.title, { color: theme.textColor }]}>
            {this.props.symbol}
          </Text>
        </View>
        <View style={styles.metaWrapper}>
        <Text style={{ color: theme.textColor }}>
          <NameText style={{ color: colors.lightGrey }}>1h</NameText>          
        </Text>
          {this._getIconPercent(this.props.percentChange1h)}
          {this._getPercentChange(this.props.percentChange1h)}
        </View>
        <View style={styles.metaWrapper}>
        <Text style={{ color: theme.textColor }}>
          <NameText style={{ color: colors.lightGrey }}>24h</NameText>          
        </Text>
          {this._getIconPercent(this.props.percentChange24h)}
          {this._getPercentChange(this.props.percentChange24h)}
        </View>
        <View style={styles.metaWrapper}>
        <Text style={{ color: theme.textColor }}>
          <NameText style={{ color: colors.lightGrey }}>7d</NameText>          
        </Text>
          {this._getIconPercent(this.props.percentChange7d)}
          {this._getPercentChange(this.props.percentChange7d)}
        </View>
        <Text style={[styles.holdingText, { color: theme.textColor }]}>
          <NameText style={[styles.holdingText, { color: colors.lightGrey }]}>
            {this.props.name}
          </NameText>
        </Text>
        <Text style={[styles.priceUsdText, { color: theme.textColor }]}>
          <NameText style={[styles.priceUsdText, { color: colors.lightGrey }]}>
            Price
          </NameText>{' '}
          {this._getPrice()}{`(${this.props.convert})`}
        </Text>
      </TouchableOpacity>
    );
  }
}

CryptoItemView.propTypes = {
  name: PropType.string,
  id: PropType.any,
  percentChange1h: PropType.number,
  priceUsd: PropType.number,
  symbol: PropType.string,
  volume24h: PropType.number,
  total: PropType.number,
  maxSupply: PropType.number,
  convert : PropType.string
}
export default CryptoItemView;