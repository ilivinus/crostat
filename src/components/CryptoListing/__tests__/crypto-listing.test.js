import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer';
import CryptoListing from '../index';
import CryptoListingPresenter from '../presenter';
import { themes } from '../../../utils/constants'

Enzyme.configure({ adapter: new Adapter() })

describe('#CryptoListing index', () => {

    let props;
    beforeEach(() => {
        const _keyExtractor = jest.fn();
        const _onEndReached = jest.fn();
        const _onRefresh = jest.fn();
        const refreshing = false;
        const _renderItem = jest.fn();
        props = {
            screenProps: {
                theme: themes
            },
            isLoading: false,
            listing: [],
            _keyExtractor,
            _onEndReached,
            _onRefresh,
            refreshing,
            _renderItem
        }
    })
    it('should render listing properly', () => {
        const component = renderer.create(<CryptoListing {...props} />).toJSON();

        expect(component).toMatchSnapshot();
    })
})

describe('#CryptoListing presenter', () => {

    let props;
    beforeEach(() => {
        props = {
            name: 'bitcoin',
            id: '2',
            percentChange1h: 23,
            priceUsd: 2,
            symbol: 'BTC',
            volume24h: 329,
            total: 13771,
            maxSupply: 379832,
            convert: 'USD',
            lastUpdate: '155322323434',
            conversion : 'USD',
            screenProps : {
                theme : themes
            }
        }

    })
    it('should render listing properly', () => {
        const component = renderer.create(<CryptoListingPresenter {...props} />).toJSON();

        expect(component).toMatchSnapshot();
    })

    
})
