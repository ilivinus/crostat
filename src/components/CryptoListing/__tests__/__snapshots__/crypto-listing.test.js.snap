// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#CryptoListing index should render listing properly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": undefined,
      },
    ]
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListFooterComponent={[Function]}
    contentContainerStyle={
      Object {
        "alignSelf": "stretch",
      }
    }
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[MockFunction]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[MockFunction]}
        refreshing={false}
        tintColor="#00C0FF"
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
      >
        <View />
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`#CryptoListing presenter should render listing properly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignSelf": "center",
      "backgroundColor": undefined,
      "flexDirection": "row",
      "height": 130,
      "opacity": 1,
      "padding": 5,
      "position": "relative",
      "width": "100%",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 15,
          "fontWeight": "400",
        },
        Array [
          Object {
            "left": 10,
            "position": "absolute",
            "top": 5,
          },
          Object {
            "color": undefined,
          },
        ],
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 14,
            "fontWeight": "500",
          },
          Array [
            Object {
              "left": 10,
              "position": "absolute",
              "top": 5,
            },
            Object {
              "color": "#B4B4B4",
            },
          ],
        ]
      }
    >
      Vol (24h)
    </Text>
     
    329
    (undefined)
  </Text>
  <Text
    style={
      Array [
        Object {
          "fontSize": 15,
          "fontWeight": "400",
        },
        Array [
          Object {
            "position": "absolute",
            "right": 10,
            "top": 5,
          },
          Object {
            "color": undefined,
          },
        ],
      ]
    }
  >
    155322323434
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0.4,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 15,
            "fontWeight": "400",
          },
          Array [
            Object {
              "fontWeight": "bold",
              "marginTop": 5,
              "textAlign": "center",
            },
            Object {
              "color": undefined,
            },
          ],
        ]
      }
    >
      BTC
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0.4,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 15,
            "fontWeight": "400",
          },
          Object {
            "color": undefined,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 14,
              "fontWeight": "500",
            },
            Object {
              "color": "#B4B4B4",
            },
          ]
        }
      >
        1h
      </Text>
    </Text>
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#4ADF86",
            "fontSize": 25,
          },
          undefined,
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      ï‰›
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontSize": 15,
            "fontWeight": "400",
          },
          Object {
            "color": "#4ADF86",
          },
        ]
      }
    >
      23.00%
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0.4,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 15,
            "fontWeight": "400",
          },
          Object {
            "color": undefined,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 14,
              "fontWeight": "500",
            },
            Object {
              "color": "#B4B4B4",
            },
          ]
        }
      >
        24h
      </Text>
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontSize": 15,
            "fontWeight": "400",
          },
          Object {
            "color": undefined,
          },
        ]
      }
    >
      ##
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0.4,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 15,
            "fontWeight": "400",
          },
          Object {
            "color": undefined,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 14,
              "fontWeight": "500",
            },
            Object {
              "color": "#B4B4B4",
            },
          ]
        }
      >
        7d
      </Text>
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontSize": 15,
            "fontWeight": "400",
          },
          Object {
            "color": undefined,
          },
        ]
      }
    >
      ##
    </Text>
  </View>
  <Text
    style={
      Array [
        Object {
          "fontSize": 15,
          "fontWeight": "400",
        },
        Array [
          Object {
            "bottom": 5,
            "left": 10,
            "position": "absolute",
          },
          Object {
            "color": undefined,
          },
        ],
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 14,
            "fontWeight": "500",
          },
          Array [
            Object {
              "bottom": 5,
              "left": 10,
              "position": "absolute",
            },
            Object {
              "color": "#B4B4B4",
            },
          ],
        ]
      }
    >
      bitcoin
    </Text>
  </Text>
  <Text
    style={
      Array [
        Object {
          "fontSize": 15,
          "fontWeight": "400",
        },
        Array [
          Object {
            "bottom": 5,
            "position": "absolute",
            "right": 10,
          },
          Object {
            "color": undefined,
          },
        ],
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 14,
            "fontWeight": "500",
          },
          Array [
            Object {
              "bottom": 5,
              "position": "absolute",
              "right": 10,
            },
            Object {
              "color": "#B4B4B4",
            },
          ],
        ]
      }
    >
      Price
    </Text>
     
    2.00000000
    (undefined)
  </Text>
</View>
`;
